In [0]:
# Class IT-533 Week #2  - Introduction to Basic Analysis with R and Python
# Summer 2020
# Lab #1 - Ramen Rating from Kaggle
# By Wei Geng(David) and Anusha Bale
# Created on: 05/25/2020
# Honor Code: “I have neither given or received, nor have I tolerated others' use of unauthorized aid.”

In [2]:
# import needed package and upload the file to google colab
from google.colab import files
uploaded = files.upload()

Saving ramen-ratings.csv to ramen-ratings (1).csv


In [3]:
# Convert to pandas DataFrame
import io
import pandas as pd
remen_df = pd.read_csv(io.BytesIO(uploaded['ramen-ratings.csv']))
remen_df.head()

Review #           Brand  ... Stars Top Ten
0      2580       New Touch  ...  3.75     NaN
1      2579        Just Way  ...     1     NaN
2      2578          Nissin  ...  2.25     NaN
3      2577         Wei Lih  ...  2.75     NaN
4      2576  Ching's Secret  ...  3.75     NaN

[5 rows x 7 columns]

### <font color = 'orange'>Question #1: What does your dataset describe?  Who collected the data and when?  For whom would an analysis be useful and interesting?</font>
<br> <font color = 'green'>Answer: This dataset was originally published on Kaggle by Aleksey Bilogur in 2018-01-11, and it contains remen rating stars and ranking information that was collected by the Ramen Rater, a product review website. This analysis would help business owner to evaluate its ramen quality and its services,but also will help the customer to shop for the best restaurants for remen noodles.
</font></br>

In [4]:
# output the data attributes and its datatypes, plus the shape of the dataset
print(remen_df.info(verbose=True))
print(remen_df.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2580 entries, 0 to 2579
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Review #  2580 non-null   int64 
 1   Brand     2580 non-null   object
 2   Variety   2580 non-null   object
 3   Style     2578 non-null   object
 4   Country   2580 non-null   object
 5   Stars     2580 non-null   object
 6   Top Ten   41 non-null     object
dtypes: int64(1), object(6)
memory usage: 141.2+ KB
None
(2580, 7)


### <font color = 'orange'>Question #2: What data does it contain, and how much?</font>
<br> <font color = 'green'>Answer: This dataset contains 7 attributes, 1 int64 column and 6 objects columns, and we need to adjust the datatype as you may notice that their data may not imported in a correct data type. In terms of number of records, we have 2,580 of them. So, this dataset contains 2,580 rows, and 7 columns.  </font>


### <font color = 'orange'>Question #3: Find the FOUR most important attributes in your dataset.  What makes them important?  What are their data types?</font>
<br> <font color = 'green'>Answer: 1) Brand 2) Style 3) Country 4)Stars are four most import attributes in the ramen dataset. They are important: 1) Brand tell us which remen noodle the rating is about 2) Style tell us the flavor of the remen 3) it tell us where the remen was produced 4) it tell us that how well the ramen noodle was rated. In terms of the datatype, we have brand = object(categorical); style = object(categorical); Country =object (categorical) and star = int64(continuous)  </font>


In [5]:
# convert nominal variable, "Brand" and "Country", to numeric, I used LableEncoder from sklearn. It is faster and easier to use.
from sklearn.preprocessing import LabelEncoder
remen_df["Brand_code"] = LabelEncoder().fit_transform(remen_df["Brand"])
remen_df["Country_code"] = LabelEncoder().fit_transform(remen_df["Country"])
remen_df.head()

Review #           Brand  ... Brand_code Country_code
0      2580       New Touch  ...        190           18
1      2579        Just Way  ...        119           32
2      2578          Nissin  ...        192           35
3      2577         Wei Lih  ...        336           32
4      2576  Ching's Secret  ...         38           16

[5 rows x 9 columns]

In [6]:
# output the basic stats for Brand_code and Country_code including mean, standard deviation, Q1 and Q3.
remen_df[['Brand_code','Country_code']].describe()

Brand_code  Country_code
count  2580.000000   2580.000000
mean    183.414341     24.159690
std      88.052819      9.976868
min       0.000000      0.000000
25%     133.000000     18.000000
50%     192.000000     29.000000
75%     242.250000     33.000000
max     354.000000     37.000000

In [7]:
# output mode for "Brand_code" and "Country_Code" 
remen_df[['Brand_code','Country_code']].mode(numeric_only=True, dropna=False, axis=0)

Brand_code  Country_code
0         192            18

In [8]:
# output median for "Brand_code" and "Country_Code" 
remen_df[['Brand_code','Country_code']].median(axis=0)

Brand_code      192.0
Country_code     29.0
dtype: float64

### <font color = 'orange'>Question #4: For TWO of these four attributes, use R to determine mean, median, mode, standard deviation, and Interquartile Range (IQR).  For the OTHER TWO of these four attributes, use Python to determine mean, median, mode, standard deviation, and Interquartile Range (IQR). If the attributes are nominal, convert them into numeric format to do this analysis.  For what purpose are these measurements important?  </font>
<br> <font color = 'green'>Answer: For the two attributes that required using R, Please check the attached R file. For the tow attributes that are required using python, I select 'Brand' and 'Country'. As you can see from the code above, we see,for 'Brand', we get mean = 183.14, median = 192, mode=192, std=88.05, q1=133, and q3=18; for 'country', we get mean=24.16, median =29, mode=18, std=9.77, q1=18, and q3=33. These measurement are very import, because they can tell you the distribution of the attributes; mean tell us the average of the values; median is the 50 percentile; mode is most frequent value; standard deviation shows how spread the data from the mean;  Q1 is the 25th percentile and Q3 is the 75th percentile.  </font>

In [0]:
# import eucliean_distance and manhattan_distance Functions
from math import*
def euclidean_distance(x,y):
    return sqrt(sum(pow(a-b,2) for a, b in zip(x, y)))
def manhattan_distance(x,y): 
    return sum(abs(a-b) for a,b in zip(x,y))

In [10]:
# The most importan attributes, I think, is 'Stars'. So, i will calculate Eucliean and Manhattan distance between, "Stars" and 'Brand' ,and "Stars" and 'Country'.  
# call the Eucliean_distance funtion and set the parameter as the first 10 datapoints of the x columns "Brand" and "Country", and y column "Stars". Return the distance value. 
import numpy as np
remen_df.loc[remen_df['Stars']=='Unrated','Stars']=np.nan
remen_df['Stars']=remen_df['Stars'].astype(float)
print("Eucliean Distance of the first 10 data points of 'Stars' and 'Brand_code' are: ",euclidean_distance(remen_df['Brand_code'][0:10],remen_df['Stars'][0:10]))
print("Eucliean Distance of the first 10 data points of 'Stars' and 'Country_code' are: ",euclidean_distance(remen_df['Country_code'][0:10],remen_df['Stars'][0:10]))

Eucliean Distance of the first 10 data points of 'Stars' and 'Brand_code' are:  579.2842027019208
Eucliean Distance of the first 10 data points of 'Stars' and 'Country_code' are:  72.7749098247466


In [11]:
# call the manhattan_distance funtion and set the parameter as the first 10 datapoints of the x columns "Brand" and "Country", and y column "Stars". Return the distance value. 
print("Manhattan Distance of the first 10 data points of 'Stars' and 'Brand_code' are: ",manhattan_distance(remen_df['Brand_code'][0:10],remen_df['Stars'][0:10]))
print("Manhattan Distance of the first 10 data points of 'Stars' and 'Country_code' are: ",manhattan_distance(remen_df['Country_code'][0:10],remen_df['Stars'][0:10]))

Manhattan Distance of the first 10 data points of 'Stars' and 'Brand_code' are:  1564.25
Manhattan Distance of the first 10 data points of 'Stars' and 'Country_code' are:  217.25


### <font color = 'orange'>Question #5: Of the four attributes with which you are working, select what you think is the most important dependent attribute (the "y" in y = fx+b) in your dataset. Then select two of the other attributes. Use Python or R to calculate the Manhattan and Euclidian distance between the first 10 attribute values of the independent and each dependent attribute. Which independent attribute is more closely related to the dependent attribute? Why does this matter?  </font>
<br> <font color = 'green'>Answer: from the results of eucliean and manhattan distances, we can conclude that 'country' are more closely related to 'Stars' than 'Brand' since 'coutry' has lower distance values. This is very important because in reality, remen noodle from one country usually has similar taste and flavors, so,  if a customer like one brand from that country, like Japan, it is likely that he would like a different brand from the same country.      </font>